### ISSUE TITLE
FPC 3 - Constructors doesn't work from script

### ISSUE THREAD

**Issue by gustaf68** [remobjects/pascalscript#154]
In FPC 3.0 and higher, calling constructors from a script doesn't work anymore. Reverting to FPC 2.6.4 (Lazarus 1.4) makes the script work again. I reported this to Lazarus bugtracker and was suggested some code in PascalScript needs to be changed:

"You should file a bug report against PascalScript. It relies on assumptions about the code generation, which we cannot support nor can guarantee to remain unchanged across releases."

There is a demo at the bugreport that shows the error: https://bugs.freepascal.org/view.php?id=31234

However, if the objects are created within the main program it works to access and use them from the script. Along these lines, at the Lazarus forum, user Derek Evans came up with a workaround that might be of help understanding how to resolve this bug: http://forum.lazarus-ide.org/index.php?topic=33302.msg215566#msg215566

(Also, and I might be deadwrong, but I got a feeling that this might be an effect of the new string handling that came with FPC 3. Its the only change I personally am aware of and that may cause this. Could be worth checking)

I have a Lazarus project that depends on PascalScript and would be grateful if someone could look at this.

**Comment by carlokok** [comment]
Is this with latest PascalScript (from this repository?) there were contributed patches for FPC3.0 that touch the constructor logic like https://github.com/remobjects/pascalscript/pull/134/files.

**Comment by gustaf68** [comment]
Hi Carlo, thanks for the reply. Yes, latest PascalScript from here. Today I tested some more, and learned that if the app is compiled as 32bits it works for both Linux and Windows. But why is there a problem with 64bit? Could it be fixed?

**Comment by gustaf68** [comment]
The crash happens in x64.inc, at CALL RAX by the red breakpoint in the picture. I don't know assembler so I can't fix it. By the way, could the ASM be replaced by pascal code instead, or does it have to be assembler? Sorry if that's a silly question. :) 

![ps 64bit bug](https://user-images.githubusercontent.com/27829191/29254012-ed44240a-808b-11e7-875f-dac4170c9edc.png)

**Comment by pult** [comment]
For x64, a replacement to "pascal" is possible. This works without failures (GPF) when an "exception" occurs ( fpc not implement asm ".params" ). Sample: https://github.com/pult/pascalscript/blob/fix_20170321_x64call/Source/x64.inc

**Comment by gustaf68** [comment]
Thanks pult, I'm checking it out. Found the enabling code:

{$DEFINE X64CALL_ASM} { optional } // TODO: Check pas version (disable "X64CALL_ASM").

After disabling X64CALL_ASM I tried compiling but got Error because the code tries to use another TRegisters that doesn't have all the members like _RSI and so on. I found the second TRegisters and will try to more tomorrow. 

Compile errors:

![ps_no_asm](https://user-images.githubusercontent.com/27829191/29254545-4c17e740-8097-11e7-9dc4-4cbf48e7f5db.png)

**Comment by gustaf68** [comment]
@pult: I'm curious why does it start with t_pcall_04 and not t_pcall_01 ?

There are very many of these procedural pointers. Maybe I'm too sleep now, but could they all be replaced by using a single open array? Oen arrays can take an unknown number of arguments. Then just loop through the array and process it:

t_pcall = procedure(RegValue: Array of IPointer);

https://www.freepascal.org/docs-html/ref/refsu68.html

**Comment by carlokok** [comment]
@gustaf68 it doesn't work that way. The asm function we use builds the stack given input parameters, an array parameter isn't compatible with that. Thats why @pult's version has all the different versions, he must have variations for all possibilities. 

The real fix would be to find out what changed in FreePascal for constructors (again for the so manyth time) and do the same change in PascalScript.

**Comment by gustaf68** [comment]
@carlokok I will try to find out the constructor change then. But then you don't think it has anything to do the new fpc string management?

**Comment by carlokok** [comment]
I don't think it has much to do with string management since the constructor you call in the referenced bugreport doesn't have string parameters.

**Comment by gustaf68** [comment]
I'm compiling old fpc revisions to find the change in code. Might take a few days, we'll see.

**Comment by gustaf68** [comment]
Found it! FPC rev. 26683 contains the changes that broke constructors in 64bit pascalscript. How do we take it from here?

------------------------------------------------------------------------
r26683 | sergei | 2014-02-06 13:26:47 +0100 (Thu, 06 Feb 2014) | 2 lines
Changed paths:
   M /trunk/compiler/ncal.pas
   M /trunk/compiler/psub.pas

* Changed passing of hidden parameters to constructors of class-style objects: VMT is always passed in the first ('self') parameter, and the second one ('vmt', although this name is now misleading) contains flag specifying the constructor behavior and meaning of 'self'. Together with r26656 this removes methodpointer sharing between two parameter node trees, it becomes possible not to put methodpointer into temp, but use node tree below 'self' parameter to do virtual calls. This also matches behavior of Delphi.
------------------------------------------------------------------------


Here's the console program I used to test revisons with.


program fptest;

{$mode objfpc}{$H+}

uses
  Crt, Classes, SysUtils, uPSRuntime, uPSComponent, uPSComponent_Default,
  Types,typinfo,rtlconsts,

  uPSC_std, uPSR_std, uPSC_classes, uPSR_classes;

var
  sl: TStringList;
  PSImport_Classes: TPSImport_Classes;
  PSScript: TPSScript;
  i:integer;
begin
  ClrScr;

  PSScript:= TPSScript.Create(nil);
  PSImport_Classes:= TPSImport_Classes.Create(nil);
  TPSPluginItem(PSScript.Plugins.Add).Plugin := PSImport_Classes;

  sl:= TStringList.Create;
  sl.Add('var');
  sl.Add('  sl: TStringList;');
  sl.Add('begin');
  sl.Add('  sl:= TStringList.Create;');
  sl.Add('  sl.Add(''Test'');');
  sl.Add('  sl.Free;');
  sl.Add('end.');

  PSScript.Script := sl;

  writeln('Attempting to compile the pascalcript...');

  if PSScript.Compile then
  begin
    writeln('Pascalscript compiled successfully!');
    if PSScript.Execute then
    begin
      writeln('Pascalscript executed successfully!');
    end
    else
    begin
      writeln('Executing the pascalscript failed!');
      Exit;
    end
  end
  else
  begin
    writeln('Compile didn''t work..');
    for i:= 0 to PSScript.CompilerMessageCount-1 do
    begin
      writeln(PSSCript.CompilerMessages[i].MessageToString);
    end;
  end;
//  readln;
end.

**Comment by carlokok** [comment]
Have a look at this:
https://github.com/remobjects/pascalscript/blob/master/Source/uPSRuntime.pas#L10259

I added that code because FPC had things in reverse; try commenting the whole ifdef FPC out.

**Comment by gustaf68** [comment]
Ok, commented out:

![screenshot_2017-08-18_13-03-51](https://user-images.githubusercontent.com/27829191/29456492-2437ede2-8416-11e7-84eb-baf9c98e263c.png)

Then ran a test app from Lazarus:

![screenshot_2017-08-18_13-01-53](https://user-images.githubusercontent.com/27829191/29456511-45ee1b96-8416-11e7-99ae-aba548a91021.png)
![screenshot_2017-08-18_13-02-26](https://user-images.githubusercontent.com/27829191/29456513-4aae9ca0-8416-11e7-975a-104b22ba98ee.png)

**Comment by gustaf68** [comment]
I'm not very good at pointers, but is there a reason Ext1 is referenced without a '^' and Ext2 is not?

```
function TPSExec.RegisterFunctionName(const Name: tbtString; ProcPtr: TPSProcPtr; Ext1, Ext2: Pointer): PProcRec;
var
  p: PProcRec;
  s: tbtString;
begin
  s := FastUppercase(Name);
  New(p);
  p^.Name := s;
  p^.Hash := MakeHash(s);
  p^.ProcPtr := ProcPtr;
  p^.FreeProc := nil;
  p.Ext1 := Ext1; 
  p^.Ext2 := Ext2;
  FRegProcs.Add(p);
  Result := P;
end;

```

**Comment by Alekcvp** [comment]
New delphi compilers automatically treat PRec.Field as PRec^.Field, so it's equal (except for with PRec^ do ...).
